File: frontend/src/lib/permissions.ts
/**
 * Mirror của {@code backend/src/main/java/com/hospital/scheduler/security/Permissions.java}.
 *
 * <p>Dùng string literal cố định (không dùng TS enum runtime) để:
 * <ul>
 *   <li>Có thể dùng câu lệnh {@code Permission.STAFF_VIEW} ở cả server và client.</li>
 *   <li>Tránh được enum runtime bị strip khi build Next.js.</li>
 * </ul>
 *
 * <p>Bất kỳ permission nào thêm mới phải được thêm vào:
 * <ol>
 *   <li>{@code backend/src/main/java/com/hospital/scheduler/security/Permissions.java}</li>
 *   <li>File này (giữ đồng bộ)</li>
 *   <li>Ma trận role-permission trong {@code DataSeeder.java} (ADMIN = all, MANAGER = all - adminOnly, STAFF = staff set)</li>
 * </ol>
 */
export const Permission = {
  // Dashboard
  DASHBOARD_VIEW: "DASHBOARD_VIEW",
  DASHBOARD_AGGREGATE: "DASHBOARD_AGGREGATE",

  // Staff
  STAFF_VIEW: "STAFF_VIEW",
  STAFF_VIEW_ALL: "STAFF_VIEW_ALL",
  STAFF_VIEW_SELF: "STAFF_VIEW_SELF",
  STAFF_CREATE: "STAFF_CREATE",
  STAFF_UPDATE: "STAFF_UPDATE",
  STAFF_DELETE: "STAFF_DELETE",
  STAFF_IMPORT: "STAFF_IMPORT",
  STAFF_EXPORT: "STAFF_EXPORT",
  STAFF_REACTIVATE: "STAFF_REACTIVATE",

  // Role / Permission matrix
  ROLE_VIEW: "ROLE_VIEW",
  ROLE_EDIT: "ROLE_EDIT",

  // Schedule period
  PERIOD_VIEW: "PERIOD_VIEW",
  PERIOD_CREATE: "PERIOD_CREATE",
  PERIOD_UPDATE: "PERIOD_UPDATE",
  PERIOD_DELETE: "PERIOD_DELETE",
  PERIOD_PUBLISH: "PERIOD_PUBLISH",
  PERIOD_ARCHIVE: "PERIOD_ARCHIVE",

  // Schedule (L01–L04)
  SCHEDULE_VIEW: "SCHEDULE_VIEW",
  SCHEDULE_CREATE: "SCHEDULE_CREATE",
  SCHEDULE_UPDATE: "SCHEDULE_UPDATE",
  SCHEDULE_DELETE: "SCHEDULE_DELETE",
  SCHEDULE_PUBLISH: "SCHEDULE_PUBLISH",
  SCHEDULE_EXPORT: "SCHEDULE_EXPORT",

  // Auto scheduling
  AUTO_SCHEDULE_VIEW: "AUTO_SCHEDULE_VIEW",
  AUTO_SCHEDULE_RUN: "AUTO_SCHEDULE_RUN",
  AUTO_SCHEDULE_APPLY: "AUTO_SCHEDULE_APPLY",
  AUTO_SCHEDULE_CONFIG_VIEW: "AUTO_SCHEDULE_CONFIG_VIEW",
  AUTO_SCHEDULE_CONFIG_EDIT: "AUTO_SCHEDULE_CONFIG_EDIT",

  // Leave
  LEAVE_VIEW: "LEAVE_VIEW",
  LEAVE_CREATE: "LEAVE_CREATE",
  LEAVE_APPROVE: "LEAVE_APPROVE",
  LEAVE_CANCEL_SELF: "LEAVE_CANCEL_SELF",

  // Exchange (swap)
  EXCHANGE_VIEW: "EXCHANGE_VIEW",
  EXCHANGE_CREATE: "EXCHANGE_CREATE",
  EXCHANGE_APPROVE: "EXCHANGE_APPROVE",
  EXCHANGE_CANCEL_SELF: "EXCHANGE_CANCEL_SELF",

  // Reports
  REPORT_VIEW: "REPORT_VIEW",
  REPORT_EXPORT: "REPORT_EXPORT",

  // Holiday
  HOLIDAY_VIEW: "HOLIDAY_VIEW",
  HOLIDAY_CREATE: "HOLIDAY_CREATE",
  HOLIDAY_UPDATE: "HOLIDAY_UPDATE",
  HOLIDAY_DELETE: "HOLIDAY_DELETE",

  // Notification
  NOTIFICATION_VIEW: "NOTIFICATION_VIEW",
  NOTIFICATION_CREATE: "NOTIFICATION_CREATE",
  NOTIFICATION_BROADCAST: "NOTIFICATION_BROADCAST",
  NOTIFICATION_MANAGE_SELF: "NOTIFICATION_MANAGE_SELF",

  // Audit / Config / Integrity
  AUDIT_VIEW: "AUDIT_VIEW",
  AUDIT_DELETE: "AUDIT_DELETE",
  APP_CONFIG_VIEW: "APP_CONFIG_VIEW",
  APP_CONFIG_EDIT: "APP_CONFIG_EDIT",
  DATA_INTEGRITY_RUN: "DATA_INTEGRITY_RUN",

  // Specialty / Shift Type / Schedule Template
  SPECIALTY_VIEW: "SPECIALTY_VIEW",
  SPECIALTY_MANAGE: "SPECIALTY_MANAGE",
  SHIFT_TYPE_MANAGE: "SHIFT_TYPE_MANAGE",
  SCHEDULE_TEMPLATE_MANAGE: "SCHEDULE_TEMPLATE_MANAGE",
} as const;

export type Permission = (typeof Permission)[keyof typeof Permission];

/**
 * Bảng mặc định role -> permission set — dùng làm fallback khi JWT không mang
 * permissions (token cũ trước khi deploy RBAC v2). Frontend vẫn nên ưu tiên
 * permissions lấy từ `useAuth().user?.permissions`.
 *
 * <p>Nguồn sự thật: backend {@code Permissions.staffPermissions()} / {@code managerPermissions()}.
 *
 * <p>Mapping theo tài liệu {@code QuanLyLichCongTac_v5.md} mục M01-F05:
 * <ul>
 *   <li><b>ADMIN</b> (Trưởng phòng): toàn quyền</li>
 *   <li><b>MANAGER</b> (Quản lý lịch): xem + phê duyệt + xếp lịch M02–M05 + auto-schedule M07</li>
 *   <li><b>STAFF</b> (Nhân viên): xem lịch cá nhân + tự đăng ký nghỉ/đổi ca</li>
 * </ul>
 */
export const RoleDefaultPermissions: Record<string, Permission[]> = {
  ADMIN: Object.values(Permission),
  MANAGER: [
    // ── Dashboard: xem ──────────────────────────────────────
    Permission.DASHBOARD_VIEW,
    Permission.DASHBOARD_AGGREGATE,

    // ── Nhân sự: xem toàn phòng (admin thêm/sửa/xoá/import) ─
    Permission.STAFF_VIEW,
    Permission.STAFF_VIEW_ALL,
    Permission.STAFF_VIEW_SELF,

    // ── Ma trận phân quyền: xem ───────────────────────────
    Permission.ROLE_VIEW,

    // ── Kỳ lịch công tác: xem (admin tạo/sửa/xoá) ────────
    Permission.PERIOD_VIEW,

    // ── Lịch trực L01–L04: xem + xếp thủ công + công bố ─
    Permission.SCHEDULE_VIEW,
    Permission.SCHEDULE_CREATE,
    Permission.SCHEDULE_UPDATE,
    Permission.SCHEDULE_DELETE,
    Permission.SCHEDULE_PUBLISH,
    Permission.SCHEDULE_EXPORT,

    // ── Specialty: xem (admin mới được CRUD) ───────────
    // BUGFIX (was SPECIALTY-MANAGER-403): /expert-clinic cần dropdown
    // chuyên khoa để filter L04. Backend cấp SPECIALTY_VIEW cho MANAGER.
    Permission.SPECIALTY_VIEW,

    // ── Auto-schedule M07: xem + chạy + áp dụng ───────────
    Permission.AUTO_SCHEDULE_VIEW,
    Permission.AUTO_SCHEDULE_RUN,
    Permission.AUTO_SCHEDULE_APPLY,
    Permission.AUTO_SCHEDULE_CONFIG_VIEW,

    // ── Nghỉ phép: xem + phê duyệt ────────────────────────
    Permission.LEAVE_VIEW,
    Permission.LEAVE_CREATE,
    Permission.LEAVE_APPROVE,
    Permission.LEAVE_CANCEL_SELF,

    // ── Đổi ca: xem + phê duyệt (M02-F04) ────────────────
    Permission.EXCHANGE_VIEW,
    Permission.EXCHANGE_CREATE,
    Permission.EXCHANGE_APPROVE,
    Permission.EXCHANGE_CANCEL_SELF,

    // ── Báo cáo: xem + xuất (M06-F04, M07-F09) ───────────
    Permission.REPORT_VIEW,
    Permission.REPORT_EXPORT,

    // ── Ngày lễ: xem (admin mới được sửa) ─────────────────
    Permission.HOLIDAY_VIEW,

    // ── Thông báo: xem (admin mới gửi broadcast) ──────────
    Permission.NOTIFICATION_VIEW,
    Permission.NOTIFICATION_MANAGE_SELF,

    // ── Audit + cấu hình: xem (admin mới sửa/xoá) ────────
    Permission.AUDIT_VIEW,
    Permission.APP_CONFIG_VIEW,
  ],
  STAFF: [
    // ── Nhân viên: xem lịch toàn phòng (read-only) + tự đăng ký ──
    // M06-F01/F02: xem lịch toàn phòng → STAFF_VIEW_ALL + SCHEDULE_VIEW
    Permission.DASHBOARD_VIEW,
    Permission.STAFF_VIEW_SELF,
    Permission.STAFF_VIEW_ALL,
    Permission.SCHEDULE_VIEW,
    // M06-F01/F02: xem kỳ lịch + lịch theo kỳ
    Permission.PERIOD_VIEW,
    Permission.HOLIDAY_VIEW,
    Permission.LEAVE_VIEW,
    Permission.LEAVE_CREATE,
    Permission.LEAVE_CANCEL_SELF,
    Permission.EXCHANGE_VIEW,
    Permission.EXCHANGE_CREATE,
    Permission.EXCHANGE_CANCEL_SELF,
    Permission.NOTIFICATION_VIEW,
    Permission.NOTIFICATION_MANAGE_SELF,
  ],
};

/**
 * Resolve the effective permission set for a user:
 *   1. If the JWT explicitly carries `permissions`, use that.
 *   2. Else fall back to the highest-privilege role's default set.
 *
 * <p>This lets the frontend degrade gracefully when a token issued before the
 * RBAC v2 rollout (no permissions claim) is still in flight.
 */
export function resolvePermissions(
  user: { roles?: string[]; permissions?: string[] } | null | undefined,
): Set<string> {
  if (!user) return new Set();
  if (Array.isArray(user.permissions) && user.permissions.length > 0) {
    return new Set(user.permissions);
  }
  const roles = user.roles ?? [];
  const priority = ["ADMIN", "MANAGER", "STAFF"];
  for (const role of priority) {
    if (roles.includes(role)) {
      return new Set(RoleDefaultPermissions[role] ?? []);
    }
  }
  return new Set();
}

File: backend/src/main/java/com/hospital/scheduler/security/Permissions.java
package com.hospital.scheduler.security;

import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Set;

/**
 * Central catalog of every permission string used by the backend.
 *
 * <p>Use these constants in {@code @PreAuthorize("hasAuthority('...')")} so that
 * the controller layer cannot drift away from the seed data. The frontend
 * mirrors this list in {@code frontend/src/lib/permissions.ts}.
 *
 * <p>Naming convention: {@code <MODULE>_<ACTION[_SCOPE]>}. New permissions
 * MUST be appended here, seeded in {@code DataSeeder}, and mirrored on the
 * frontend before being used in any controller.
 */
public final class Permissions {

    // ── Dashboard ───────────────────────────────────────────────────────
    public static final String DASHBOARD_VIEW = "DASHBOARD_VIEW";
    public static final String DASHBOARD_AGGREGATE = "DASHBOARD_AGGREGATE";

    // ── Staff ───────────────────────────────────────────────────────────
    public static final String STAFF_VIEW = "STAFF_VIEW";
    public static final String STAFF_CREATE = "STAFF_CREATE";
    public static final String STAFF_UPDATE = "STAFF_UPDATE";
    public static final String STAFF_DELETE = "STAFF_DELETE";
    public static final String STAFF_IMPORT = "STAFF_IMPORT";

    // ── Role / Permission matrix ────────────────────────────────────────
    public static final String ROLE_VIEW = "ROLE_VIEW";
    public static final String ROLE_EDIT = "ROLE_EDIT";

    // ── Schedule period ─────────────────────────────────────────────────
    public static final String PERIOD_VIEW = "PERIOD_VIEW";
    public static final String PERIOD_CREATE = "PERIOD_CREATE";
    public static final String PERIOD_UPDATE = "PERIOD_UPDATE";
    public static final String PERIOD_DELETE = "PERIOD_DELETE";
    public static final String PERIOD_PUBLISH = "PERIOD_PUBLISH";
    public static final String PERIOD_ARCHIVE = "PERIOD_ARCHIVE";

    // ── Schedule (L01–L04) ──────────────────────────────────────────────
    public static final String SCHEDULE_VIEW = "SCHEDULE_VIEW";
    public static final String SCHEDULE_CREATE = "SCHEDULE_CREATE";
    public static final String SCHEDULE_UPDATE = "SCHEDULE_UPDATE";
    public static final String SCHEDULE_DELETE = "SCHEDULE_DELETE";
    public static final String SCHEDULE_PUBLISH = "SCHEDULE_PUBLISH";
    public static final String SCHEDULE_EXPORT = "SCHEDULE_EXPORT";

    // ── Auto scheduling (M07) ───────────────────────────────────────────
    public static final String AUTO_SCHEDULE_VIEW = "AUTO_SCHEDULE_VIEW";
    public static final String AUTO_SCHEDULE_RUN = "AUTO_SCHEDULE_RUN";
    public static final String AUTO_SCHEDULE_APPLY = "AUTO_SCHEDULE_APPLY";
    public static final String AUTO_SCHEDULE_CONFIG_VIEW = "AUTO_SCHEDULE_CONFIG_VIEW";
    public static final String AUTO_SCHEDULE_CONFIG_EDIT = "AUTO_SCHEDULE_CONFIG_EDIT";

    // ── Leave ───────────────────────────────────────────────────────────
    public static final String LEAVE_VIEW = "LEAVE_VIEW";
    public static final String LEAVE_CREATE = "LEAVE_CREATE";
    public static final String LEAVE_APPROVE = "LEAVE_APPROVE";
    public static final String LEAVE_CANCEL_SELF = "LEAVE_CANCEL_SELF";

    // ── Schedule exchange / swap ────────────────────────────────────────
    public static final String EXCHANGE_VIEW = "EXCHANGE_VIEW";
    public static final String EXCHANGE_CREATE = "EXCHANGE_CREATE";
    public static final String EXCHANGE_APPROVE = "EXCHANGE_APPROVE";
    public static final String EXCHANGE_CANCEL_SELF = "EXCHANGE_CANCEL_SELF";

    // ── Reports / statistics ────────────────────────────────────────────
    public static final String REPORT_VIEW = "REPORT_VIEW";
    public static final String REPORT_EXPORT = "REPORT_EXPORT";

    // ── Holiday ─────────────────────────────────────────────────────────
    public static final String HOLIDAY_VIEW = "HOLIDAY_VIEW";
    public static final String HOLIDAY_CREATE = "HOLIDAY_CREATE";
    public static final String HOLIDAY_UPDATE = "HOLIDAY_UPDATE";
    public static final String HOLIDAY_DELETE = "HOLIDAY_DELETE";

    // ── Notification ────────────────────────────────────────────────────
    public static final String NOTIFICATION_VIEW = "NOTIFICATION_VIEW";
    public static final String NOTIFICATION_CREATE = "NOTIFICATION_CREATE";
    public static final String NOTIFICATION_BROADCAST = "NOTIFICATION_BROADCAST";
    public static final String NOTIFICATION_MANAGE_SELF = "NOTIFICATION_MANAGE_SELF";

    // ── Audit / System log / Config / Integrity ─────────────────────────
    public static final String AUDIT_VIEW = "AUDIT_VIEW";
    public static final String AUDIT_DELETE = "AUDIT_DELETE";
    public static final String SYSTEM_LOG_VIEW = "SYSTEM_LOG_VIEW";
    public static final String APP_CONFIG_VIEW = "APP_CONFIG_VIEW";
    public static final String APP_CONFIG_EDIT = "APP_CONFIG_EDIT";
    public static final String DATA_INTEGRITY_RUN = "DATA_INTEGRITY_RUN";

    // ── Specialty / Shift Type ──────────────────────────────────────────
    public static final String SPECIALTY_MANAGE = "SPECIALTY_MANAGE";
    public static final String SHIFT_TYPE_MANAGE = "SHIFT_TYPE_MANAGE";
    public static final String SCHEDULE_TEMPLATE_MANAGE = "SCHEDULE_TEMPLATE_MANAGE";

    private Permissions() {
        // utility class
    }

    /**
     * Ordered catalog used by the seeder so the order in the database is stable.
     */
    public static Map<String, String> catalog() {
        Map<String, String> m = new LinkedHashMap<>();
        m.put(DASHBOARD_VIEW,          "Xem dashboard tổng quan");
        m.put(DASHBOARD_AGGREGATE,     "Xem chỉ số tổng hợp toàn hệ thống");

        m.put(STAFF_VIEW,              "Xem thông tin nhân sự");
        m.put(STAFF_CREATE,            "Tạo nhân sự mới");
        m.put(STAFF_UPDATE,            "Cập nhật thông tin nhân sự");
        m.put(STAFF_DELETE,            "Xóa / vô hiệu hóa nhân sự");
        m.put(STAFF_IMPORT,            "Import nhân sự từ Excel/CSV");

        m.put(ROLE_VIEW,               "Xem ma trận phân quyền");
        m.put(ROLE_EDIT,               "Sửa ma trận phân quyền");

        m.put(PERIOD_VIEW,             "Xem kỳ lịch công tác");
        m.put(PERIOD_CREATE,           "Tạo kỳ lịch công tác");
        m.put(PERIOD_UPDATE,           "Cập nhật kỳ lịch công tác");
        m.put(PERIOD_DELETE,           "Xóa kỳ lịch công tác");
        m.put(PERIOD_PUBLISH,          "Công bố kỳ lịch công tác");
        m.put(PERIOD_ARCHIVE,          "Lưu trữ kỳ lịch công tác");

        m.put(SCHEDULE_VIEW,           "Xem lịch trực (cá nhân/toàn hệ thống)");
        m.put(SCHEDULE_CREATE,         "Tạo lịch trực");
        m.put(SCHEDULE_UPDATE,         "Cập nhật lịch trực");
        m.put(SCHEDULE_DELETE,         "Xóa lịch trực");
        m.put(SCHEDULE_PUBLISH,        "Công bố lịch trực hàng loạt");
        m.put(SCHEDULE_EXPORT,         "Xuất lịch trực ra Excel/PDF");

        m.put(AUTO_SCHEDULE_VIEW,            "Xem phương án auto-schedule");
        m.put(AUTO_SCHEDULE_RUN,             "Chạy auto-schedule");
        m.put(AUTO_SCHEDULE_APPLY,           "Áp dụng phương án auto-schedule");
        m.put(AUTO_SCHEDULE_CONFIG_VIEW,     "Xem cấu hình auto-schedule");
        m.put(AUTO_SCHEDULE_CONFIG_EDIT,     "Sửa cấu hình auto-schedule");

        m.put(LEAVE_VIEW,              "Xem yêu cầu nghỉ phép");
        m.put(LEAVE_CREATE,            "Tạo yêu cầu nghỉ phép");
        m.put(LEAVE_APPROVE,           "Duyệt / từ chối yêu cầu nghỉ phép");
        m.put(LEAVE_CANCEL_SELF,       "Hủy yêu cầu nghỉ phép của chính mình");

        m.put(EXCHANGE_VIEW,           "Xem yêu cầu đổi ca");
        m.put(EXCHANGE_CREATE,         "Tạo yêu cầu đổi ca");
        m.put(EXCHANGE_APPROVE,        "Duyệt / từ chối yêu cầu đổi ca");
        m.put(EXCHANGE_CANCEL_SELF,    "Hủy yêu cầu đổi ca của chính mình");

        m.put(REPORT_VIEW,             "Xem báo cáo thống kê");
        m.put(REPORT_EXPORT,           "Xuất báo cáo");

        m.put(HOLIDAY_VIEW,            "Xem ngày lễ");
        m.put(HOLIDAY_CREATE,          "Tạo ngày lễ");
        m.put(HOLIDAY_UPDATE,          "Cập nhật ngày lễ");
        m.put(HOLIDAY_DELETE,          "Xóa ngày lễ");

        m.put(NOTIFICATION_VIEW,          "Xem thông báo");
        m.put(NOTIFICATION_CREATE,        "Tạo thông báo");
        m.put(NOTIFICATION_BROADCAST,     "Gửi thông báo broadcast tới nhiều người");
        m.put(NOTIFICATION_MANAGE_SELF,   "Quản lý thông báo của chính mình (đánh dấu đã đọc)");

        m.put(AUDIT_VIEW,              "Xem lịch sử thao tác");
        m.put(AUDIT_DELETE,            "Xóa lịch sử thao tác");
        m.put(SYSTEM_LOG_VIEW,         "Xem system log");
        m.put(APP_CONFIG_VIEW,         "Xem cấu hình hệ thống");
        m.put(APP_CONFIG_EDIT,         "Sửa cấu hình hệ thống");
        m.put(DATA_INTEGRITY_RUN,      "Chạy kiểm tra tính toàn vẹn dữ liệu");

        m.put(SPECIALTY_MANAGE,        "Quản lý danh mục chuyên khoa");
        m.put(SHIFT_TYPE_MANAGE,       "Quản lý danh mục loại ca");
        m.put(SCHEDULE_TEMPLATE_MANAGE, "Quản lý mẫu lịch trực");

        return m;
    }

    /**
     * Permissions granted to ADMIN (everything).
     */
    public static Set<String> allPermissions() {
        return catalog().keySet();
    }

    /**
     * Permissions reserved for ADMIN only — MANAGER is denied even though it
     * has nearly everything else.
     */
    public static Set<String> adminOnlyPermissions() {
        return Set.of(
                ROLE_EDIT,
                AUDIT_DELETE,
                SYSTEM_LOG_VIEW,
                DATA_INTEGRITY_RUN,
                AUTO_SCHEDULE_CONFIG_EDIT,
                APP_CONFIG_EDIT
        );
    }

    /**
     * Permissions granted to MANAGER (everything except {@link #adminOnlyPermissions()}).
     */
    public static Set<String> managerPermissions() {
        java.util.Set<String> all = new java.util.LinkedHashSet<>(allPermissions());
        all.removeAll(adminOnlyPermissions());
        return all;
    }

    /**
     * Permissions granted to STAFF — self-service only.
     */
    public static Set<String> staffPermissions() {
        return Set.of(
                DASHBOARD_VIEW,
                STAFF_VIEW,
                SCHEDULE_VIEW,
                HOLIDAY_VIEW,
                LEAVE_VIEW,
                LEAVE_CREATE,
                LEAVE_CANCEL_SELF,
                EXCHANGE_VIEW,
                EXCHANGE_CREATE,
                EXCHANGE_CANCEL_SELF,
                NOTIFICATION_VIEW,
                NOTIFICATION_MANAGE_SELF
        );
    }
}